src/map.cpp

```diff
@@ -5652,6 +5656,79 @@ void map::draw( const catacurses::window &w, const tripoint &center )
     }
 }
 
+void map::update_terrain_weather()
+{
+    const w_point weatherPoint = *g->weather_precise;
+    if( g->weather == WEATHER_FLURRIES ) {
+        snowfall += 2;
+    } else if( g->weather == WEATHER_SNOW ) {
+        snowfall += 4;
+    } else if( g->weather == WEATHER_SNOWSTORM ) {
+        snowfall += 8;
+    } else {
+        snowfall -= std::max( 0, ( static_cast<int>( weatherPoint.temperature - 32 ) ) );
+    }
+    snowfall = std::max( 0, snowfall );
+    snowfall = std::min( 172800, snowfall );
+}
+
+void map::add_snow()
+{
+    if( g->weather >= WEATHER_FLURRIES && snowfall > 3000 ) {
+        tripoint gp( 0, 0, 0 );
+        int total_grid = 0;
+        int snowfall_settled = 0;
+        int &gx = gp.x;
+        int &gy = gp.y;
+        int snowfactor = std::min( 15000, snowfall );
+        snowfactor = 15001 - snowfactor;
+        for( gx = 0; gx < my_MAPSIZE * SEEX; ++gx ) {
+            for( gy = 0; gy < my_MAPSIZE * SEEY; ++gy ) {
+                total_grid += 1;
+                const auto grid_offset = point {gx, gy};
+                const tripoint map_location = tripoint( grid_offset, 0 );
+                if( ( !ter( map_location )->has_flag( TFLAG_INDOORS ) ) &&
+                    ( !ter( map_location )->has_flag( "TREE" ) ) && ( is_outside( map_location ) ) &&
+                    ( furn( map_location ) == f_null ) ) {
+                    if( one_in( std::max( 2, static_cast<int>( snowfactor / 2 ) ) ) ) {
+                        furn_set( map_location, f_snow );
+                        snowfall_settled += 1;
+                    }
+                }
+            }
+        }
+        percent_settled = std::min( 100, static_cast<int>( ( snowfall_settled / total_grid ) * 100 ) );
```

Comment:
1. `snowfall_settled` increments only when there was no snow on the tile initially and it was randomly placed there. Which means it denotes the amount of snow placed per invocation of `add_snow`, not the actual ratio. This seems to be wrong.

2. `snowfall_settled / total_grid` isn't what you would expect. Both are integers and produce either `0` or `1` (nothing else in the current implementation). Should be `static_cast<float>(snowfall_settled) / total_grid` to produce decimals.

3. `std::min( 100, ...)` is excessive, the ratio never exceeds `1`.